## Instructions
To move forward with the project, you need to create a collection of songs with their audio features - as large as possible! <br>
<br>
These are the songs that we will cluster. And, later, when the user inputs a song, we will find the cluster to which the song belongs and recommend a song from the same cluster. The more songs you have, the more accurate and diverse recommendations you'll be able to give. Although... you might want to make sure the collected songs are "curated" in a certain way. Try to find playlists of songs that are diverse, but also that meet certain standards. <br>
<br>
The process of sending hundreds or thousands of requests can take some time - it's normal if you have to wait a few minutes (or, if you're ambitious, even hours) to get all the data you need.<br>
<br>
An idea for collecting as many songs as possible is to start with all the songs of a big, diverse playlist and then go to every artist present in the playlist and grab every song of every album of that artist. The amount of songs you'll be collecting per playlist will grow exponentially!

### Import libraries

In [1]:
import pandas as pd
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
from getpass import getpass
from random import sample

### Initialize SpotiPy with user credentials

In [2]:
client_id="4617eb96ef9545239b68c0d443d5baba"
client_secret="93d966e07836444e9f4a9ddd881f0060"

In [3]:
# client_id = getpass("Insert client ID: ")
# client_secret = getpass("Insert client secret: ")

In [4]:
sp = spotipy.Spotify(auth_manager=SpotifyClientCredentials(client_id, client_secret))

### Load playlist

In [5]:
# Check the name of the playlist

playlist = sp.playlist("37i9dQZF1DWUbj5dNZ89An")
playlist

{'collaborative': False,
 'description': 'Canciones para disfrutar del viaje',
 'external_urls': {'spotify': 'https://open.spotify.com/playlist/37i9dQZF1DWUbj5dNZ89An'},
 'followers': {'href': None, 'total': 420811},
 'href': 'https://api.spotify.com/v1/playlists/37i9dQZF1DWUbj5dNZ89An?additional_types=track',
 'id': '37i9dQZF1DWUbj5dNZ89An',
 'images': [{'height': None,
   'url': 'https://i.scdn.co/image/ab67706f0000000306a77b3763f413baed261b23',
   'width': None}],
 'name': 'Canta en el coche',
 'owner': {'display_name': 'Spotify',
  'external_urls': {'spotify': 'https://open.spotify.com/user/spotify'},
  'href': 'https://api.spotify.com/v1/users/spotify',
  'id': 'spotify',
  'type': 'user',
  'uri': 'spotify:user:spotify'},
 'primary_color': '#ffffff',
 'public': False,
 'snapshot_id': 'MCxiZjlkODBhYWNmMDA3MGI5ZWIxOGQ5YjIzNGRkYzdmZjM0YzZkOTEz',
 'tracks': {'href': 'https://api.spotify.com/v1/playlists/37i9dQZF1DWUbj5dNZ89An/tracks?offset=0&limit=100&additional_types=track',
  'item

In [6]:
playlist["name"]

'Canta en el coche'

### Obtain the name and id of the tracks

In [7]:
playlist = sp.user_playlist_tracks("spotify", "37i9dQZF1DWUbj5dNZ89An")
playlist.keys()

dict_keys(['href', 'items', 'limit', 'next', 'offset', 'previous', 'total'])

In [8]:
playlist["items"][0].keys()

dict_keys(['added_at', 'added_by', 'is_local', 'primary_color', 'track', 'video_thumbnail'])

In [9]:
playlist["items"][0]["track"].keys()

dict_keys(['album', 'artists', 'available_markets', 'disc_number', 'duration_ms', 'episode', 'explicit', 'external_ids', 'external_urls', 'href', 'id', 'is_local', 'name', 'popularity', 'preview_url', 'track', 'track_number', 'type', 'uri'])

In [10]:
playlist["items"][0]["track"]["name"]

'La Bebe - Remix'

In [11]:
playlist["items"][0]["track"]["id"]

'2UW7JaomAMuX9pZrjVpHAU'

In [12]:
artists = []
song_artists = []

for artist in playlist["items"][99]["track"]["artists"]:
    song_artists.append(artist["name"])

joined_song_artists = ", ".join(song_artists)
artists.append(joined_song_artists)
artists

['Aitana, Emilia, Ptazeta']

In [13]:
# Function to obtain all the tracks titles, ids and artists from a playlist created by Spotify

def get_track_id_title(playlist_id):
    
    titles = []
    ids = []
    artists = []

    playlist = sp.user_playlist_tracks("spotify", playlist_id)
    for i in range(0, len(playlist["items"])):
        song_artists = []
        title = playlist["items"][i]["track"]["name"]
        song_id = playlist["items"][i]["track"]["id"]
        for artist in playlist["items"][i]["track"]["artists"]:
            song_artists.append(artist["name"])
        titles.append(title)
        ids.append(song_id)
        joined_song_artists = ", ".join(song_artists)
        artists.append(joined_song_artists)
        
    return titles, ids, artists

In [14]:
titles_canta_coche, ids_canta_coche, artists_canta_coche = get_track_id_title("37i9dQZF1DWUbj5dNZ89An")

In [15]:
sample(titles_canta_coche, 10)

['Manos Frías (con Reik, Beéle)',
 'TUYA',
 "I Ain't Worried",
 'Numb',
 'All My Life (feat. J. Cole)',
 'Flaca',
 'Quieres',
 'Fuera del mercado',
 'Destino o casualidad',
 'DESPECHÁ']

In [16]:
sample(ids_canta_coche, 10)

['4HzqWEdAbzcVFDv0pvxV4w',
 '0Sfn2TYbpQtCGMBf6C0Y6T',
 '3USxtqRwSYz57Ewm6wWRMp',
 '20CozgjF6bshBw8cLhN23B',
 '3iHzKA9HlXf5wsGdsrsnSA',
 '23E53f3C9SJ9MqBh1f6hEw',
 '6otePxalBK8AVa20xhZYVQ',
 '1MB8kTH7VKvAMfL9SHgJmG',
 '4rQu72zFJz3qne1bnEkVSt',
 '0FDTPGlLjF8SGWSsHyzNBe']

In [17]:
sample(artists_canta_coche, 10)

['Camilo, Shawn Mendes',
 'Abraham Mateo, Ana Mena',
 'Sky Rompiendo, Feid, Myke Towers',
 'Pablo Alborán, David Bisbal, Antonio Carmona, Jesse & Joy, Juanes, Malú, Vanesa Martín, India Martinez, Antonio Orozco, Niña Pastori, Laura Pausini, Abel Pintos, Rozalén, Shakira, Tommy Torres, Manuel Carrasco, Pablo López',
 'sangiovanni, Aitana',
 'Metro Boomin, Swae Lee, NAV, A Boogie Wit da Hoodie',
 'Taylor Swift',
 'Bizarrap, Shakira',
 'KAROL G',
 'Sebastian Yatra, Manuel Turizo, Beéle']

### Obtain audio features

In [18]:
sp.audio_features(ids_canta_coche)

[{'danceability': 0.812,
  'energy': 0.479,
  'key': 2,
  'loudness': -5.678,
  'mode': 0,
  'speechiness': 0.333,
  'acousticness': 0.213,
  'instrumentalness': 1.21e-06,
  'liveness': 0.0756,
  'valence': 0.559,
  'tempo': 169.922,
  'type': 'audio_features',
  'id': '2UW7JaomAMuX9pZrjVpHAU',
  'uri': 'spotify:track:2UW7JaomAMuX9pZrjVpHAU',
  'track_href': 'https://api.spotify.com/v1/tracks/2UW7JaomAMuX9pZrjVpHAU',
  'analysis_url': 'https://api.spotify.com/v1/audio-analysis/2UW7JaomAMuX9pZrjVpHAU',
  'duration_ms': 234353,
  'time_signature': 4},
 {'danceability': 0.652,
  'energy': 0.8,
  'key': 9,
  'loudness': -4.019,
  'mode': 0,
  'speechiness': 0.0614,
  'acousticness': 0.143,
  'instrumentalness': 0.629,
  'liveness': 0.112,
  'valence': 0.234,
  'tempo': 143.978,
  'type': 'audio_features',
  'id': '7ro0hRteUMfnOioTFI5TG1',
  'uri': 'spotify:track:7ro0hRteUMfnOioTFI5TG1',
  'track_href': 'https://api.spotify.com/v1/tracks/7ro0hRteUMfnOioTFI5TG1',
  'analysis_url': 'https://a

### Create dataframe

In [19]:
canta_coche = pd.json_normalize(sp.audio_features(ids_canta_coche))
canta_coche["title"] = titles_canta_coche 
canta_coche["artist"] = artists_canta_coche 
canta_coche

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.812   0.479    2    -5.678     0       0.3330        0.2130   
1          0.652   0.800    9    -4.019     0       0.0614        0.1430   
2          0.748   0.761    2    -6.621     1       0.1020        0.1220   
3          0.748   0.856   11    -3.517     0       0.0348        0.0824   
4          0.783   0.592   11    -8.332     0       0.0331        0.4350   
..           ...     ...  ...       ...   ...          ...           ...   
95         0.481   0.810    9    -5.704     1       0.1330        0.2340   
96         0.705   0.757    6    -5.608     0       0.0656        0.1950   
97         0.711   0.727    2    -3.150     1       0.1620        0.1680   
98         0.768   0.714   10    -5.110     1       0.0401        0.3520   
99         0.796   0.646   11    -3.925     0       0.1580        0.0744   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000001    0.0756    0.559  169.922  audio_features   
1           0.629000    0.1120    0.234  143.978  audio_features   
2           0.000007    0.0984    0.362  116.041  audio_features   
3           0.000000    0.1420    0.927  123.031  audio_features   
4           0.000004    0.3470    0.726  120.018  audio_features   
..               ...       ...      ...      ...             ...   
95          0.000000    0.0809    0.328  175.697  audio_features   
96          0.000228    0.0524    0.737  104.973  audio_features   
97          0.000000    0.2960    0.465  180.039  audio_features   
98          0.000000    0.1500    0.842  131.443  audio_features   
99          0.000000    0.1270    0.663  180.030  audio_features   

                        id                                   uri  \
0   2UW7JaomAMuX9pZrjVpHAU  spotify:track:2UW7JaomAMuX9pZrjVpHAU   
1   7ro0hRteUMfnOioTFI5TG1  spotify:track:7ro0hRteUMfnOioTFI5TG1   
2   5TMLavqGRKNxgzMBrVpC9R  spotify:track:5TMLavqGRKNxgzMBrVpC9R   
3   0Be7sopyKMv8Y8npsUkax2  spotify:track:0Be7sopyKMv8Y8npsUkax2   
4   7FbrGaHYVDmfr7KoLIZnQ7  spotify:track:7FbrGaHYVDmfr7KoLIZnQ7   
..                     ...                                   ...   
95  0mGvaKpwLYoybnIoboC3uS  spotify:track:0mGvaKpwLYoybnIoboC3uS   
96  3SK45LddxlEkzI8OWO9Eyo  spotify:track:3SK45LddxlEkzI8OWO9Eyo   
97  2V4hc5gplg7EXlrRUvD4rp  spotify:track:2V4hc5gplg7EXlrRUvD4rp   
98  5YqEzk3C5c3UZ1D5fJUlXA  spotify:track:5YqEzk3C5c3UZ1D5fJUlXA   
99  4iGnpp1h0i33lpoP42Y8Jt  spotify:track:4iGnpp1h0i33lpoP42Y8Jt   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/2UW7JaomAMuX...   
1   https://api.spotify.com/v1/tracks/7ro0hRteUMfn...   
2   https://api.spotify.com/v1/tracks/5TMLavqGRKNx...   
3   https://api.spotify.com/v1/tracks/0Be7sopyKMv8...   
4   https://api.spotify.com/v1/tracks/7FbrGaHYVDmf...   
..                                                ...   
95  https://api.spotify.com/v1/tracks/0mGvaKpwLYoy...   
96  https://api.spotify.com/v1/tracks/3SK45LddxlEk...   
97  https://api.spotify.com/v1/tracks/2V4hc5gplg7E...   
98  https://api.spotify.com/v1/tracks/5YqEzk3C5c3U...   
99  https://api.spotify.com/v1/tracks/4iGnpp1h0i33...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/2UW7...       234353   
1   https://api.spotify.com/v1/audio-analysis/7ro0...       231704   
2   https://api.spotify.com/v1/audio-analysis/5TML...       179940   
3   https://api.spotify.com/v1/audio-analysis/0Be7...       189320   
4   https://api.spotify.com/v1/audio-analysis/7Fbr...       174253   
..                                                ...          ...   
95  https://api.spotify.com/v1/audio-analysis/0mGv...       232253   
96  https://api.spotify.com/v1/audio-analysis/3SK4...       187053   
97  https://api.spotify.com/v1/audio-analysis/2V4h...       174380   
98  https://api.spotify.com/v1/audio-analysis/5YqE...       163026   
99

### Repeat with more playlists

In [20]:
# Function to obtain audio features and create dataframe

def df_audio_features(playlist_id):
    titles, ids, artists = get_track_id_title(playlist_id)
    audio_features = sp.audio_features(ids)
    df = pd.json_normalize(audio_features)
    df["title"] = titles
    df["artist"] = artists
    return df

In [21]:
df_list = [canta_coche]

In [22]:
country_hits = df_audio_features("37i9dQZF1DWZBCPUIUs2iR")
df_list.append(country_hits)
country_hits

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.554   0.783    6    -5.852     1       0.0457        0.4160   
1          0.614   0.680    6    -4.580     1       0.0289        0.3690   
2          0.708   0.524    5   -12.644     1       0.0352        0.0551   
3          0.657   0.639   10    -7.527     1       0.0412        0.5810   
4          0.671   0.501    4   -13.119     1       0.0594        0.3420   
..           ...     ...  ...       ...   ...          ...           ...   
95         0.739   0.717    3    -9.568     1       0.0308        0.4170   
96         0.259   0.590    3    -8.343     1       0.0363        0.4720   
97         0.596   0.626    7   -12.945     1       0.0634        0.4020   
98         0.472   0.338    2   -11.196     1       0.0273        0.8500   
99         0.373   0.562    2   -12.841     1       0.0323        0.2160   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000002    0.6310    0.813  105.390  audio_features   
1           0.000002    0.1150    0.707  149.959  audio_features   
2           0.000003    0.0371    0.690  111.482  audio_features   
3           0.000000    0.2780    0.674  124.882  audio_features   
4           0.000000    0.1940    0.860   87.040  audio_features   
..               ...       ...      ...      ...             ...   
95          0.000918    0.3340    0.663  113.234  audio_features   
96          0.001280    0.1350    0.474  201.016  audio_features   
97          0.000000    0.0498    0.962  155.556  audio_features   
98          0.000221    0.0761    0.522   85.660  audio_features   
99          0.103000    0.3700    0.708  103.109  audio_features   

                        id                                   uri  \
0   4w3tQBXhn5345eUXDGBWZG  spotify:track:4w3tQBXhn5345eUXDGBWZG   
1   6foY66mWZN0pSRjZ408c00  spotify:track:6foY66mWZN0pSRjZ408c00   
2   6iy4PoAuZBMvtrlDX4VxC7  spotify:track:6iy4PoAuZBMvtrlDX4VxC7   
3   6sxptembJVty4sNtcPMAVz  spotify:track:6sxptembJVty4sNtcPMAVz   
4   5KqldkCunQ2rWxruMEtGh0  spotify:track:5KqldkCunQ2rWxruMEtGh0   
..                     ...                                   ...   
95  70TsRbycWfabtevnripyRu  spotify:track:70TsRbycWfabtevnripyRu   
96  4ON9pkEiboV83FUHkeIxGp  spotify:track:4ON9pkEiboV83FUHkeIxGp   
97  5EPQbU0wk70Btiu9onZR48  spotify:track:5EPQbU0wk70Btiu9onZR48   
98  4NRFiS0pHjTFDjOm1LWTZj  spotify:track:4NRFiS0pHjTFDjOm1LWTZj   
99  1B3wswoDyVdEbWO9lmEzSk  spotify:track:1B3wswoDyVdEbWO9lmEzSk   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/4w3tQBXhn534...   
1   https://api.spotify.com/v1/tracks/6foY66mWZN0p...   
2   https://api.spotify.com/v1/tracks/6iy4PoAuZBMv...   
3   https://api.spotify.com/v1/tracks/6sxptembJVty...   
4   https://api.spotify.com/v1/tracks/5KqldkCunQ2r...   
..                                                ...   
95  https://api.spotify.com/v1/tracks/70TsRbycWfab...   
96  https://api.spotify.com/v1/tracks/4ON9pkEiboV8...   
97  https://api.spotify.com/v1/tracks/5EPQbU0wk70B...   
98  https://api.spotify.com/v1/tracks/4NRFiS0pHjTF...   
99  https://api.spotify.com/v1/tracks/1B3wswoDyVdE...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/4w3t...       162467   
1   https://api.spotify.com/v1/audio-analysis/6foY...       234347   
2   https://api.spotify.com/v1/audio-analysis/6iy4...       200667   
3   https://api.spotify.com/v1/audio-analysis/6sxp...       234493   
4   https://api.spotify.com/v1/audio-analysis/5Kql...       214880   
..                                                ...          ...   
95  https://api.spotify.com/v1/audio-analysis/70Ts...       223533   
96  https://api.spotify.com/v1/audio-analysis/4ON9...       235920   
97  https://api.spotify.com/v1/audio-analysis/5EPQ...       177693   
98  https://api.spotify.com/v1/audio-analysis/4NRF...       174827   
99

In [23]:
iconic_soundtracks = df_audio_features("37i9dQZF1DX1tz6EDao8it")
df_list.append(iconic_soundtracks)
iconic_soundtracks

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.4650  0.6870   10   -10.964     1       0.0495        0.0206   
1         0.2860  0.0968    7   -16.806     1       0.0323        0.1780   
2         0.1440  0.1320   10   -19.115     1       0.0333        0.9460   
3         0.1800  0.2260    9   -16.457     0       0.0420        0.9510   
4         0.4830  0.1920    4   -15.378     0       0.0469        0.9910   
..           ...     ...  ...       ...   ...          ...           ...   
77        0.0938  0.1580    9   -18.539     1       0.0495        0.9490   
78        0.1390  0.2210    0   -16.382     1       0.0374        0.7610   
79        0.1750  0.1710    4   -19.589     0       0.0411        0.9700   
80        0.2520  0.1440   10   -17.657     1       0.0283        0.7580   
81        0.2790  0.1020    2   -19.933     1       0.0365        0.8520   

    instrumentalness  liveness  valence    tempo            type  \
0              0.929    0.1210   0.1300   70.507  audio_features   
1              0.696    0.0861   0.0395   62.881  audio_features   
2              0.779    0.1150   0.0627  102.283  audio_features   
3              0.982    0.0998   0.0480   94.079  audio_features   
4              0.908    0.1060   0.0542  100.848  audio_features   
..               ...       ...      ...      ...             ...   
77             0.921    0.1110   0.0347   86.925  audio_features   
78             0.892    0.3180   0.0630   78.841  audio_features   
79             0.900    0.2080   0.0746  122.278  audio_features   
80             0.901    0.0852   0.0677  182.776  audio_features   
81             0.904    0.2160   0.0599  138.581  audio_features   

                        id                                   uri  \
0   08QaHlMPWuO5PUxjl61bXn  spotify:track:08QaHlMPWuO5PUxjl61bXn   
1   6ZFbXIJkuI1dVNWvzJzown  spotify:track:6ZFbXIJkuI1dVNWvzJzown   
2   2TZbQZXOuR8osP2AK8yYMN  spotify:track:2TZbQZXOuR8osP2AK8yYMN   
3   6pWgRkpqVfxnj3WuIcJ7WP  spotify:track:6pWgRkpqVfxnj3WuIcJ7WP   
4   14rZjW3RioG7WesZhYESso  spotify:track:14rZjW3RioG7WesZhYESso   
..                     ...                                   ...   
77  5aS4hzWTQ968fPIoKsDg1q  spotify:track:5aS4hzWTQ968fPIoKsDg1q   
78  7aVRZwGYPypqF7AMN0pwCG  spotify:track:7aVRZwGYPypqF7AMN0pwCG   
79  1CxT3WZkSuEAxPOKAPzDDl  spotify:track:1CxT3WZkSuEAxPOKAPzDDl   
80  6yzzno5cXBkh5c6UzRxcyM  spotify:track:6yzzno5cXBkh5c6UzRxcyM   
81  1X26LGm0cnHhzAvXpIYIWc  spotify:track:1X26LGm0cnHhzAvXpIYIWc   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/08QaHlMPWuO5...   
1   https://api.spotify.com/v1/tracks/6ZFbXIJkuI1d...   
2   https://api.spotify.com/v1/tracks/2TZbQZXOuR8o...   
3   https://api.spotify.com/v1/tracks/6pWgRkpqVfxn...   
4   https://api.spotify.com/v1/tracks/14rZjW3RioG7...   
..                                                ...   
77  https://api.spotify.com/v1/tracks/5aS4hzWTQ968...   
78  https://api.spotify.com/v1/tracks/7aVRZwGYPypq...   
79  https://api.spotify.com/v1/tracks/1CxT3WZkSuEA...   
80  https://api.spotify.com/v1/tracks/6yzzno5cXBkh...   
81  https://api.spotify.com/v1/tracks/1X26LGm0cnHh...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/08Qa...        90427   
1   https://api.spotify.com/v1/audio-analysis/6ZFb...       275556   
2   https://api.spotify.com/v1/audio-analysis/2TZb...       207427   
3   https://api.spotify.com/v1/audio-analysis/6pWg...       126960   
4   https://api.spotify.com/v1/audio-analysis/14rZ...       140733   
..                                                ...          ...   
77  https://api.spotify.com/v1/audio-analysis/5aS4...       235587   
78  https://api.spotify.com/v1/audio-analysis/7aVR...       244133   
79  https://api.spotify.com/v1/audio-analysis/1CxT...       128320   
80  https://api.spotify.com/v1/audio-analysis/6yzz...       310267   
81

In [24]:
samba_classics = df_audio_features("37i9dQZF1DX2OwDeeyTh0V")
df_list.append(samba_classics)
samba_classics

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.674   0.419    7   -12.413     0       0.2680        0.7470   
1          0.596   0.451    5    -9.468     1       0.0404        0.6400   
2          0.605   0.662   10    -6.993     1       0.0924        0.4320   
3          0.695   0.924    9    -4.178     1       0.0454        0.2290   
4          0.705   0.337   11   -14.457     1       0.3430        0.2840   
..           ...     ...  ...       ...   ...          ...           ...   
70         0.793   0.420    1    -8.817     1       0.1380        0.1310   
71         0.672   0.786    9    -9.381     0       0.2560        0.7990   
72         0.652   0.535    9   -12.210     1       0.0386        0.9080   
73         0.742   0.655    2    -9.615     1       0.0696        0.8720   
74         0.793   0.617   10    -8.953     0       0.1070        0.0732   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000000    0.0797    0.870  150.131  audio_features   
1           0.000004    0.1580    0.486  148.045  audio_features   
2           0.000000    0.1160    0.953   93.771  audio_features   
3           0.000000    0.2640    0.964   94.827  audio_features   
4           0.000003    0.0747    0.600   96.556  audio_features   
..               ...       ...      ...      ...             ...   
70          0.001360    0.1140    0.721  104.175  audio_features   
71          0.004320    0.1020    0.888  123.648  audio_features   
72          0.000053    0.1170    0.961   99.097  audio_features   
73          0.000182    0.1420    0.845  113.951  audio_features   
74          0.000087    0.1410    0.323  112.001  audio_features   

                        id                                   uri  \
0   7fs6csqE3EltCrESv6cn0J  spotify:track:7fs6csqE3EltCrESv6cn0J   
1   5Ikos3uiBfPYAf7EZfz8t3  spotify:track:5Ikos3uiBfPYAf7EZfz8t3   
2   0m3hiKjI2QGSKxvG732Ekq  spotify:track:0m3hiKjI2QGSKxvG732Ekq   
3   0rImThhPOV8wtgeyltD6no  spotify:track:0rImThhPOV8wtgeyltD6no   
4   3C8ULtjiUmTtcpE7QJBZ7G  spotify:track:3C8ULtjiUmTtcpE7QJBZ7G   
..                     ...                                   ...   
70  7f1jaIAFffaCl5Hs49HZAG  spotify:track:7f1jaIAFffaCl5Hs49HZAG   
71  25HD49CToA0SUAdsz2QdtF  spotify:track:25HD49CToA0SUAdsz2QdtF   
72  3TBy6We8pwdcLeFNt0ahKA  spotify:track:3TBy6We8pwdcLeFNt0ahKA   
73  7Fpa5YcWRvaYfwNyGjnAqQ  spotify:track:7Fpa5YcWRvaYfwNyGjnAqQ   
74  76gxg0ZNq873zRgVHEAIR9  spotify:track:76gxg0ZNq873zRgVHEAIR9   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/7fs6csqE3Elt...   
1   https://api.spotify.com/v1/tracks/5Ikos3uiBfPY...   
2   https://api.spotify.com/v1/tracks/0m3hiKjI2QGS...   
3   https://api.spotify.com/v1/tracks/0rImThhPOV8w...   
4   https://api.spotify.com/v1/tracks/3C8ULtjiUmTt...   
..                                                ...   
70  https://api.spotify.com/v1/tracks/7f1jaIAFffaC...   
71  https://api.spotify.com/v1/tracks/25HD49CToA0S...   
72  https://api.spotify.com/v1/tracks/3TBy6We8pwdc...   
73  https://api.spotify.com/v1/tracks/7Fpa5YcWRvaY...   
74  https://api.spotify.com/v1/tracks/76gxg0ZNq873...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/7fs6...       144760   
1   https://api.spotify.com/v1/audio-analysis/5Iko...       228133   
2   https://api.spotify.com/v1/audio-analysis/0m3h...       162360   
3   https://api.spotify.com/v1/audio-analysis/0rIm...       232800   
4   https://api.spotify.com/v1/audio-analysis/3C8U...       205200   
..                                                ...          ...   
70  https://api.spotify.com/v1/audio-analysis/7f1j...       159427   
71  https://api.spotify.com/v1/audio-analysis/25HD...       185813   
72  https://api.spotify.com/v1/audio-analysis/3TBy...       139267   
73  https://api.spotify.com/v1/audio-analysis/7Fpa...       194427   
74

In [25]:
feelin_good = df_audio_features("37i9dQZF1DX9XIFQuFvzM4")
df_list.append(feelin_good)
feelin_good

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.663   0.600    7   -10.870     1       0.0320        0.4300   
1          0.527   0.415    4   -11.451     0       0.1220        0.4570   
2          0.650   0.306    9    -9.443     1       0.0393        0.5700   
3          0.572   0.418    0   -10.738     1       0.0349        0.6350   
4          0.769   0.367    2   -11.226     1       0.0312        0.6840   
..           ...     ...  ...       ...   ...          ...           ...   
95         0.565   0.644    6    -8.116     1       0.0696        0.4260   
96         0.703   0.529    7   -11.112     1       0.1000        0.0168   
97         0.713   0.800    7    -5.938     0       0.0349        0.0491   
98         0.761   0.103   10   -16.964     1       0.0416        0.9440   
99         0.623   0.824    6    -8.971     0       0.0479        0.0195   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000000    0.1840    0.800  129.991  audio_features   
1           0.000017    0.1170    0.515   78.169  audio_features   
2           0.000007    0.0707    0.605  118.068  audio_features   
3           0.000000    0.0961    0.694  104.566  audio_features   
4           0.000016    0.0810    0.535  103.621  audio_features   
..               ...       ...      ...      ...             ...   
95          0.000000    0.2640    0.633   97.716  audio_features   
96          0.000570    0.9110    0.772  128.059  audio_features   
97          0.000080    0.0921    0.814  139.942  audio_features   
98          0.000000    0.1270    0.530  114.044  audio_features   
99          0.130000    0.0491    0.574  111.356  audio_features   

                        id                                   uri  \
0   7tqhbajSfrz2F7E1Z75ASX  spotify:track:7tqhbajSfrz2F7E1Z75ASX   
1   1k1Bqnv2R0uJXQN4u6LKYt  spotify:track:1k1Bqnv2R0uJXQN4u6LKYt   
2   3SdTKo2uVsxFblQjpScoHy  spotify:track:3SdTKo2uVsxFblQjpScoHy   
3   745H5CctFr12Mo7cqa1BMH  spotify:track:745H5CctFr12Mo7cqa1BMH   
4   3zBhihYUHBmGd2bcQIobrF  spotify:track:3zBhihYUHBmGd2bcQIobrF   
..                     ...                                   ...   
95  5kQ2ZEav7TgUoLSLrm7h8S  spotify:track:5kQ2ZEav7TgUoLSLrm7h8S   
96  1MQWtVcs0PKsY4PA6ZvLiy  spotify:track:1MQWtVcs0PKsY4PA6ZvLiy   
97  3zYQ1KXzqBWRsxIqdVMoYF  spotify:track:3zYQ1KXzqBWRsxIqdVMoYF   
98  373yGkNo74RZgmQgTiR8xK  spotify:track:373yGkNo74RZgmQgTiR8xK   
99  6igsoAR6Co9u7Rq3U7mlOD  spotify:track:6igsoAR6Co9u7Rq3U7mlOD   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/7tqhbajSfrz2...   
1   https://api.spotify.com/v1/tracks/1k1Bqnv2R0uJ...   
2   https://api.spotify.com/v1/tracks/3SdTKo2uVsxF...   
3   https://api.spotify.com/v1/tracks/745H5CctFr12...   
4   https://api.spotify.com/v1/tracks/3zBhihYUHBmG...   
..                                                ...   
95  https://api.spotify.com/v1/tracks/5kQ2ZEav7TgU...   
96  https://api.spotify.com/v1/tracks/1MQWtVcs0PKs...   
97  https://api.spotify.com/v1/tracks/3zYQ1KXzqBWR...   
98  https://api.spotify.com/v1/tracks/373yGkNo74RZ...   
99  https://api.spotify.com/v1/tracks/6igsoAR6Co9u...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/7tqh...       151667   
1   https://api.spotify.com/v1/audio-analysis/1k1B...       125093   
2   https://api.spotify.com/v1/audio-analysis/3SdT...       180056   
3   https://api.spotify.com/v1/audio-analysis/745H...       165000   
4   https://api.spotify.com/v1/audio-analysis/3zBh...       163756   
..                                                ...          ...   
95  https://api.spotify.com/v1/audio-analysis/5kQ2...       452773   
96  https://api.spotify.com/v1/audio-analysis/1MQW...       179427   
97  https://api.spotify.com/v1/audio-analysis/3zYQ...       177200   
98  https://api.spotify.com/v1/audio-analysis/373y...       160133   
99

In [26]:
summer_jazz = df_audio_features("37i9dQZF1DWTKxc7ZObqeH")
df_list.append(summer_jazz)
summer_jazz

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.557   0.229    7   -15.943     1       0.0359         0.853   
1          0.510   0.189    6   -15.775     1       0.0341         0.972   
2          0.634   0.232    3   -18.569     1       0.0373         0.970   
3          0.555   0.240    0   -15.647     1       0.0348         0.977   
4          0.433   0.267    5   -15.967     1       0.0388         0.973   
..           ...     ...  ...       ...   ...          ...           ...   
95         0.502   0.216    8   -18.930     1       0.0322         0.943   
96         0.455   0.112    9   -20.048     1       0.0371         0.898   
97         0.470   0.174    7   -18.095     1       0.0316         0.895   
98         0.527   0.214   10   -17.607     1       0.0342         0.764   
99         0.561   0.200    4   -18.437     0       0.0325         0.934   

    instrumentalness  liveness  valence    tempo            type  \
0              0.423    0.0984   0.0816   61.858  audio_features   
1              0.928    0.1000   0.2160  103.792  audio_features   
2              0.902    0.1050   0.3310  122.381  audio_features   
3              0.839    0.1380   0.3940  127.822  audio_features   
4              0.896    0.1030   0.0928  119.322  audio_features   
..               ...       ...      ...      ...             ...   
95             0.958    0.1170   0.1000  100.908  audio_features   
96             0.866    0.1260   0.3330   62.585  audio_features   
97             0.934    0.1040   0.3070   87.750  audio_features   
98             0.417    0.0741   0.1760   80.566  audio_features   
99             0.847    0.1710   0.3730  105.699  audio_features   

                        id                                   uri  \
0   1bB1538FOBYVUs2z06ssQ3  spotify:track:1bB1538FOBYVUs2z06ssQ3   
1   2nUhKiUfVdLcGmcdYHdjBw  spotify:track:2nUhKiUfVdLcGmcdYHdjBw   
2   1H460eor4CG8p8OAz7TGYM  spotify:track:1H460eor4CG8p8OAz7TGYM   
3   2xVVCXzIuILeIRoQrZYlQD  spotify:track:2xVVCXzIuILeIRoQrZYlQD   
4   2mKGI5TnGVSWbloCZlma6B  spotify:track:2mKGI5TnGVSWbloCZlma6B   
..                     ...                                   ...   
95  5FCu0Dq8UHHEHmbvA3zsxv  spotify:track:5FCu0Dq8UHHEHmbvA3zsxv   
96  5CHYbV9TiI3V823hZBqN92  spotify:track:5CHYbV9TiI3V823hZBqN92   
97  4q86fROSuJgs5J8pX8YBZM  spotify:track:4q86fROSuJgs5J8pX8YBZM   
98  5ctxuHXM6daOwGJYUxrwp3  spotify:track:5ctxuHXM6daOwGJYUxrwp3   
99  3XEbmnM9sO4aAmwzlZoBca  spotify:track:3XEbmnM9sO4aAmwzlZoBca   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/1bB1538FOBYV...   
1   https://api.spotify.com/v1/tracks/2nUhKiUfVdLc...   
2   https://api.spotify.com/v1/tracks/1H460eor4CG8...   
3   https://api.spotify.com/v1/tracks/2xVVCXzIuILe...   
4   https://api.spotify.com/v1/tracks/2mKGI5TnGVSW...   
..                                                ...   
95  https://api.spotify.com/v1/tracks/5FCu0Dq8UHHE...   
96  https://api.spotify.com/v1/tracks/5CHYbV9TiI3V...   
97  https://api.spotify.com/v1/tracks/4q86fROSuJgs...   
98  https://api.spotify.com/v1/tracks/5ctxuHXM6daO...   
99  https://api.spotify.com/v1/tracks/3XEbmnM9sO4a...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/1bB1...       258873   
1   https://api.spotify.com/v1/audio-analysis/2nUh...       169863   
2   https://api.spotify.com/v1/audio-analysis/1H46...       220410   
3   https://api.spotify.com/v1/audio-analysis/2xVV...       208096   
4   https://api.spotify.com/v1/audio-analysis/2mKG...       191843   
..                                                ...          ...   
95  https://api.spotify.com/v1/audio-analysis/5FCu...       216778   
96  https://api.spotify.com/v1/audio-analysis/5CHY...       177353   
97  https://api.spotify.com/v1/audio-analysis/4q86...       206457   
98  https://api.spotify.com/v1/audio-analysis/5ctx...       160414   
99

In [27]:
sad_songs = df_audio_features("37i9dQZF1DX7qK8ma5wgG1")
df_list.append(sad_songs)
sad_songs

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.508  0.4300    2    -9.475     0       0.0335         0.830   
1          0.445  0.2320    1    -9.145     1       0.0294         0.886   
2          0.676  0.4610    3    -3.590     1       0.0304         0.555   
3          0.554  0.3260    9    -7.085     1       0.0299         0.697   
4          0.637  0.5130    1    -6.797     1       0.0367         0.518   
..           ...     ...  ...       ...   ...          ...           ...   
75         0.785  0.3180    7    -7.794     0       0.0316         0.866   
76         0.824  0.3000    1    -6.678     0       0.0565         0.854   
77         0.691  0.0682    5   -18.186     1       0.0645         0.951   
78         0.405  0.2790   10    -8.099     1       0.0334         0.881   
79         0.561  0.4310   10    -8.810     1       0.0578         0.768   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000441    0.0930    0.324  130.090  audio_features   
1           0.000003    0.1280    0.246   72.005  audio_features   
2           0.000000    0.0951    0.466  107.950  audio_features   
3           0.000000    0.0876    0.278   71.267  audio_features   
4           0.000000    0.1240    0.234  141.907  audio_features   
..               ...       ...      ...      ...             ...   
75          0.000000    0.1150    0.424  120.117  audio_features   
76          0.000000    0.1740    0.227  112.867  audio_features   
77          0.000000    0.0916    0.352   76.940  audio_features   
78          0.000000    0.0933    0.256  119.850  audio_features   
79          0.000014    0.1060    0.137  143.875  audio_features   

                        id                                   uri  \
0   1odExI7RdWc4BT515LTAwj  spotify:track:1odExI7RdWc4BT515LTAwj   
1   0V9QY6NnXtAMwjltgMP0pl  spotify:track:0V9QY6NnXtAMwjltgMP0pl   
2   3gdPwk2wyOXNRnTA1KXnEr  spotify:track:3gdPwk2wyOXNRnTA1KXnEr   
3   7djOJmSvxXtq0Q49iC9qsX  spotify:track:7djOJmSvxXtq0Q49iC9qsX   
4   22UYp7aAy1XmIYZ1GsnJuB  spotify:track:22UYp7aAy1XmIYZ1GsnJuB   
..                     ...                                   ...   
75  1ayOMuDEULRqZ4QYVwSNV7  spotify:track:1ayOMuDEULRqZ4QYVwSNV7   
76  1cpVLszZnKUHUjsJDTEi3D  spotify:track:1cpVLszZnKUHUjsJDTEi3D   
77  6hjh2phISvfR1Ib6S0FK7b  spotify:track:6hjh2phISvfR1Ib6S0FK7b   
78  21dpBHPTV7ythCHBI6Slcw  spotify:track:21dpBHPTV7ythCHBI6Slcw   
79  5wANPM4fQCJwkGd4rN57mH  spotify:track:5wANPM4fQCJwkGd4rN57mH   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/1odExI7RdWc4...   
1   https://api.spotify.com/v1/tracks/0V9QY6NnXtAM...   
2   https://api.spotify.com/v1/tracks/3gdPwk2wyOXN...   
3   https://api.spotify.com/v1/tracks/7djOJmSvxXtq...   
4   https://api.spotify.com/v1/tracks/22UYp7aAy1Xm...   
..                                                ...   
75  https://api.spotify.com/v1/tracks/1ayOMuDEULRq...   
76  https://api.spotify.com/v1/tracks/1cpVLszZnKUH...   
77  https://api.spotify.com/v1/tracks/6hjh2phISvfR...   
78  https://api.spotify.com/v1/tracks/21dpBHPTV7yt...   
79  https://api.spotify.com/v1/tracks/5wANPM4fQCJw...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/1odE...       212954   
1   https://api.spotify.com/v1/audio-analysis/0V9Q...       204503   
2   https://api.spotify.com/v1/audio-analysis/3gdP...       210880   
3   https://api.spotify.com/v1/audio-analysis/7djO...       227925   
4   https://api.spotify.com/v1/audio-analysis/22UY...       202193   
..                                                ...          ...   
75  https://api.spotify.com/v1/audio-analysis/1ayO...       163000   
76  https://api.spotify.com/v1/audio-analysis/1cpV...       175980   
77  https://api.spotify.com/v1/audio-analysis/6hjh...       222468   
78  https://api.spotify.com/v1/audio-analysis/21dp...       303653   
79

In [28]:
lofi_cafe = df_audio_features("37i9dQZF1DX9RwfGbeGQwP")
df_list.append(lofi_cafe)
lofi_cafe

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.746  0.2490    9   -16.632     1       0.0539       0.88200   
1          0.684  0.3070    1   -11.707     1       0.0514       0.53600   
2          0.585  0.3570    7   -13.928     0       0.0730       0.84600   
3          0.698  0.7010    0    -8.358     0       0.0465       0.07360   
4          0.677  0.3500   11   -10.977     0       0.0601       0.65000   
..           ...     ...  ...       ...   ...          ...           ...   
95         0.766  0.2230    0   -16.046     1       0.0749       0.59900   
96         0.615  0.1940    4   -17.496     0       0.0512       0.90800   
97         0.682  0.5510   10    -7.629     1       0.2130       0.76600   
98         0.571  0.4230    8    -9.896     0       0.0340       0.00831   
99         0.586  0.0979    0   -19.274     0       0.0324       0.64000   

    instrumentalness  liveness  valence    tempo            type  \
0              0.925    0.1120    0.261  120.043  audio_features   
1              0.960    0.1070    0.766   81.986  audio_features   
2              0.891    0.1140    0.505  161.420  audio_features   
3              0.787    0.3670    0.602   89.993  audio_features   
4              0.901    0.2050    0.623   82.008  audio_features   
..               ...       ...      ...      ...             ...   
95             0.922    0.0563    0.355   84.563  audio_features   
96             0.866    0.1660    0.118   79.991  audio_features   
97             0.762    0.6180    0.655   84.955  audio_features   
98             0.863    0.8920    0.625  174.022  audio_features   
99             0.924    0.1270    0.111   80.991  audio_features   

                        id                                   uri  \
0   6zkW8DxvN281ruHccv0MIS  spotify:track:6zkW8DxvN281ruHccv0MIS   
1   0YexmBFxPG2pLlSTBmiTvW  spotify:track:0YexmBFxPG2pLlSTBmiTvW   
2   7BAfo8WKIrAm3hJDj16Xd6  spotify:track:7BAfo8WKIrAm3hJDj16Xd6   
3   3dnJ9V91tzRy6QjwyagZtu  spotify:track:3dnJ9V91tzRy6QjwyagZtu   
4   2uhlAgdxgzwMvJ1YEA5ZaT  spotify:track:2uhlAgdxgzwMvJ1YEA5ZaT   
..                     ...                                   ...   
95  3fRppxMcSzxXIvQwEuMtV4  spotify:track:3fRppxMcSzxXIvQwEuMtV4   
96  2dysxhiq58BpAlFIUYd0aj  spotify:track:2dysxhiq58BpAlFIUYd0aj   
97  7McgbJW6Bo2WHx9tXmzoUA  spotify:track:7McgbJW6Bo2WHx9tXmzoUA   
98  3Rwjlw8iBHUH8Kd7SLFQeX  spotify:track:3Rwjlw8iBHUH8Kd7SLFQeX   
99  1cfqOZQ4RDSf13bU3YiiHn  spotify:track:1cfqOZQ4RDSf13bU3YiiHn   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/6zkW8DxvN281...   
1   https://api.spotify.com/v1/tracks/0YexmBFxPG2p...   
2   https://api.spotify.com/v1/tracks/7BAfo8WKIrAm...   
3   https://api.spotify.com/v1/tracks/3dnJ9V91tzRy...   
4   https://api.spotify.com/v1/tracks/2uhlAgdxgzwM...   
..                                                ...   
95  https://api.spotify.com/v1/tracks/3fRppxMcSzxX...   
96  https://api.spotify.com/v1/tracks/2dysxhiq58Bp...   
97  https://api.spotify.com/v1/tracks/7McgbJW6Bo2W...   
98  https://api.spotify.com/v1/tracks/3Rwjlw8iBHUH...   
99  https://api.spotify.com/v1/tracks/1cfqOZQ4RDSf...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/6zkW...       129868   
1   https://api.spotify.com/v1/audio-analysis/0Yex...       123659   
2   https://api.spotify.com/v1/audio-analysis/7BAf...       144000   
3   https://api.spotify.com/v1/audio-analysis/3dnJ...       110885   
4   https://api.spotify.com/v1/audio-analysis/2uhl...       121447   
..                                                ...          ...   
95  https://api.spotify.com/v1/audio-analysis/3fRp...       124869   
96  https://api.spotify.com/v1/audio-analysis/2dys...       144000   
97  https://api.spotify.com/v1/audio-analysis/7Mcg...       140118   
98  https://api.spotify.com/v1/audio-analysis/3Rwj...       191873   
99

In [48]:
happy_hits = df_audio_features("37i9dQZF1DXdPec7aLTmlC")
df_list.append(happy_hits)
happy_hits

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.561   0.965    7    -3.673     0       0.0343       0.00383   
1          0.704   0.797    0    -5.927     1       0.0475       0.08260   
2          0.514   0.730    1    -5.934     1       0.0598       0.00146   
3          0.671   0.845   11    -4.930     0       0.0480       0.02070   
4          0.548   0.816    0    -4.209     1       0.0465       0.12200   
..           ...     ...  ...       ...   ...          ...           ...   
95         0.464   0.724    7    -5.019     1       0.0705       0.01880   
96         0.643   0.585    2    -5.994     1       0.0369       0.22500   
97         0.862   0.753    5    -5.356     1       0.0625       0.13100   
98         0.763   0.661    4    -5.592     0       0.0560       0.16900   
99         0.722   0.893    0    -3.255     0       0.0475       0.04410   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000007    0.3710    0.304  128.040  audio_features   
1           0.000745    0.0546    0.825  139.994  audio_features   
2           0.000095    0.0897    0.334  171.005  audio_features   
3           0.000000    0.3290    0.775  110.056  audio_features   
4           0.000000    0.3350    0.557   95.390  audio_features   
..               ...       ...      ...      ...             ...   
95          0.000000    0.6790    0.303   82.526  audio_features   
96          0.000000    0.0927    0.188  106.015  audio_features   
97          0.000002    0.0770    0.827  105.039  audio_features   
98          0.000000    0.1170    0.632   95.002  audio_features   
99          0.000937    0.0673    0.439  119.932  audio_features   

                        id                                   uri  \
0   4uUG5RXrOk84mYEfFvj3cK  spotify:track:4uUG5RXrOk84mYEfFvj3cK   
1   4h9wh7iOZ0GGn8QVp4RAOB  spotify:track:4h9wh7iOZ0GGn8QVp4RAOB   
2   0VjIjW4GlUZAMYd2vXMi3b  spotify:track:0VjIjW4GlUZAMYd2vXMi3b   
3   1vYXt7VSjH9JIM5oRRo7vA  spotify:track:1vYXt7VSjH9JIM5oRRo7vA   
4   6UelLqGlWMcVH1E5c4H7lY  spotify:track:6UelLqGlWMcVH1E5c4H7lY   
..                     ...                                   ...   
95  0rc1HCVoReqzzXF9jssqZk  spotify:track:0rc1HCVoReqzzXF9jssqZk   
96  21H0pUV2uPIPcFnFhgTBjo  spotify:track:21H0pUV2uPIPcFnFhgTBjo   
97  7MC4XR9M9amdmKQr2iDF7i  spotify:track:7MC4XR9M9amdmKQr2iDF7i   
98  2SUxn2O9NHL6GHGQFgwCY0  spotify:track:2SUxn2O9NHL6GHGQFgwCY0   
99  3iw6V4LH7yPj1ESORX9RIN  spotify:track:3iw6V4LH7yPj1ESORX9RIN   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/4uUG5RXrOk84...   
1   https://api.spotify.com/v1/tracks/4h9wh7iOZ0GG...   
2   https://api.spotify.com/v1/tracks/0VjIjW4GlUZA...   
3   https://api.spotify.com/v1/tracks/1vYXt7VSjH9J...   
4   https://api.spotify.com/v1/tracks/6UelLqGlWMcV...   
..                                                ...   
95  https://api.spotify.com/v1/tracks/0rc1HCVoReqz...   
96  https://api.spotify.com/v1/tracks/21H0pUV2uPIP...   
97  https://api.spotify.com/v1/tracks/7MC4XR9M9amd...   
98  https://api.spotify.com/v1/tracks/2SUxn2O9NHL6...   
99  https://api.spotify.com/v1/tracks/3iw6V4LH7yPj...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/4uUG...       175238   
1   https://api.spotify.com/v1/audio-analysis/4h9w...       148486   
2   https://api.spotify.com/v1/audio-analysis/0VjI...       200040   
3   https://api.spotify.com/v1/audio-analysis/1vYX...       176579   
4   https://api.spotify.com/v1/audio-analysis/6Uel...       174000   
..                                                ...          ...   
95  https://api.spotify.com/v1/audio-analysis/0rc1...       174749   
96  https://api.spotify.com/v1/audio-analysis/21H0...       216873   
97  https://api.spotify.com/v1/audio-analysis/7MC4...       182857   
98  https://api.spotify.com/v1/audio-analysis/2SUx...       153627   
99

In [49]:
classical_romance = df_audio_features("37i9dQZF1DX4s3V2rTswzO")
df_list.append(classical_romance)
classical_romance

danceability   energy  key  loudness  mode  speechiness  acousticness  \
0         0.4430  0.05210   10   -22.964     1       0.0492         0.992   
1         0.0921  0.01830    7   -31.858     1       0.0513         0.942   
2         0.4290  0.04720   11   -27.076     1       0.0870         0.995   
3         0.2380  0.00484    8   -31.760     1       0.0393         0.995   
4         0.0791  0.05720    7   -29.864     1       0.0494         0.890   
5         0.4060  0.00921    2   -31.822     1       0.0870         0.984   
6         0.2270  0.00482    8   -36.351     1       0.0360         0.993   
7         0.2070  0.00805    9   -36.380     1       0.0366         0.995   
8         0.4390  0.02510    4   -29.029     1       0.0381         0.994   
9         0.2130  0.08680   10   -27.444     1       0.0413         0.952   
10        0.2690  0.00229    1   -39.158     1       0.0372         0.986   
11        0.3270  0.00319    3   -32.189     1       0.0355         0.994   
12        0.3710  0.02470    1   -26.056     1       0.0550         0.993   
13        0.1830  0.00750    0   -33.087     1       0.0395         0.974   
14        0.2800  0.15400    0   -25.846     1       0.0439         0.989   
15        0.1110  0.05080   10   -26.315     1       0.0464         0.962   
16        0.4410  0.00957    7   -26.839     1       0.0308         0.987   
17        0.5000  0.01040    2   -26.764     1       0.0622         0.996   
18        0.1630  0.03660    7   -27.021     1       0.0451         0.980   
19        0.1690  0.01610    9   -31.034     0       0.0442         0.986   
20        0.4120  0.01450    8   -30.382     1       0.0481         0.995   
21        0.4350  0.00889    7   -33.986     1       0.0460         0.992   
22        0.1650  0.02470    2   -23.367     1       0.0446         0.984   
23        0.3060  0.00253    1   -42.185     1       0.0401         0.982   
24        0.1960  0.04960    6   -24.536     0       0.0415         0.990   
25        0.4060  0.00645    7   -32.739     0       0.0453         0.994   
26        0.3420  0.00369    7   -32.610     1       0.0509         0.994   
27        0.2990  0.00626    9   -30.923     1       0.0478         0.994   
28        0.2960  0.00896    7   -31.948     0       0.0394         0.994   
29        0.2530  0.02720    5   -31.271     1       0.0431         0.993   
30        0.3680  0.00363    1   -36.759     1       0.0350         0.993   
31        0.1790  0.04310    0   -23.664     0       0.0465         0.987   
32        0.1380  0.03370    2   -27.266     1       0.0476         0.989   
33        0.2290  0.04120    3   -26.158     1       0.0467         0.965   
34        0.3900  0.04290    3   -27.186     1       0.0371         0.966   
35        0.2930  0.06770    3   -24.559     1       0.0472         0.981   
36        0.1680  0.06130    4   -24.987     0       0.0481         0.974   
37        0.3770  0.01050    6   -32.994     1       0.0426         0.994   
38        0.2680  0.07820    7   -29.935     1       0.0415         0.993   
39        0.2320  0.01170    7   -33.034     1       0.0395         0.976   
40        0.3280  0.00278   10   -38.801     1       0.0566         0.995   
41        0.3260  0.00502   11   -28.664     1       0.0383         0.993   
42        0.2730  0.00735    9   -32.622     1       0.0375         0.991   
43        0.1760  0.04160    4   -26.636     1       0.0402         0.968   
44        0.4380  0.01560    3   -29.108     1       0.0621         0.996   
45        0.2300  0.03610    0   -28.250     0       0.0482         0.983   
46        0.3690  0.01540    4   -27.040     0       0.0468         0.992   
47        0.1780  0.17300   10   -16.399     1       0.0404         0.941   
48        0.4000  0.01150    2   -39.280     1       0.0488         0.994   
49        0.2370  0.24700    1   -14.859     1       0.0338         0.878   
50        0.1690  0.05240    9   -29.896     1       0.0744         0.937   


### Join the dataframes

In [50]:
audio_features_songs = pd.concat(df_list).reset_index(drop=True)
audio_features_songs

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0           0.812  0.4790    2    -5.678     0       0.3330        0.2130   
1           0.652  0.8000    9    -4.019     0       0.0614        0.1430   
2           0.748  0.7610    2    -6.621     1       0.1020        0.1220   
3           0.748  0.8560   11    -3.517     0       0.0348        0.0824   
4           0.783  0.5920   11    -8.332     0       0.0331        0.4350   
..            ...     ...  ...       ...   ...          ...           ...   
883         0.369  0.0154    4   -27.040     0       0.0468        0.9920   
884         0.178  0.1730   10   -16.399     1       0.0404        0.9410   
885         0.400  0.0115    2   -39.280     1       0.0488        0.9940   
886         0.237  0.2470    1   -14.859     1       0.0338        0.8780   
887         0.169  0.0524    9   -29.896     1       0.0744        0.9370   

     instrumentalness  liveness  valence    tempo            type  \
0            0.000001    0.0756    0.559  169.922  audio_features   
1            0.629000    0.1120    0.234  143.978  audio_features   
2            0.000007    0.0984    0.362  116.041  audio_features   
3            0.000000    0.1420    0.927  123.031  audio_features   
4            0.000004    0.3470    0.726  120.018  audio_features   
..                ...       ...      ...      ...             ...   
883          0.224000    0.1040    0.173   72.766  audio_features   
884          0.884000    0.1270    0.153   74.681  audio_features   
885          0.930000    0.0715    0.392  125.504  audio_features   
886          0.942000    0.0939    0.311   93.520  audio_features   
887          0.000556    0.0590    0.037   72.665  audio_features   

                         id                                   uri  \
0    2UW7JaomAMuX9pZrjVpHAU  spotify:track:2UW7JaomAMuX9pZrjVpHAU   
1    7ro0hRteUMfnOioTFI5TG1  spotify:track:7ro0hRteUMfnOioTFI5TG1   
2    5TMLavqGRKNxgzMBrVpC9R  spotify:track:5TMLavqGRKNxgzMBrVpC9R   
3    0Be7sopyKMv8Y8npsUkax2  spotify:track:0Be7sopyKMv8Y8npsUkax2   
4    7FbrGaHYVDmfr7KoLIZnQ7  spotify:track:7FbrGaHYVDmfr7KoLIZnQ7   
..                      ...                                   ...   
883  4RrqS5QeuCvi4VY4a8CDmZ  spotify:track:4RrqS5QeuCvi4VY4a8CDmZ   
884  492xOpyuUiVZ74qhKc2e2w  spotify:track:492xOpyuUiVZ74qhKc2e2w   
885  0y6TO3w7SWXOeIHKw4eps7  spotify:track:0y6TO3w7SWXOeIHKw4eps7   
886  3o9Yhvfzz2M7vpznPrCmFi  spotify:track:3o9Yhvfzz2M7vpznPrCmFi   
887  4ghT12J2hsyfn0JZ4RDWrf  spotify:track:4ghT12J2hsyfn0JZ4RDWrf   

                                            track_href  \
0    https://api.spotify.com/v1/tracks/2UW7JaomAMuX...   
1    https://api.spotify.com/v1/tracks/7ro0hRteUMfn...   
2    https://api.spotify.com/v1/tracks/5TMLavqGRKNx...   
3    https://api.spotify.com/v1/tracks/0Be7sopyKMv8...   
4    https://api.spotify.com/v1/tracks/7FbrGaHYVDmf...   
..                                                 ...   
883  https://api.spotify.com/v1/tracks/4RrqS5QeuCvi...   
884  https://api.spotify.com/v1/tracks/492xOpyuUiVZ...   
885  https://api.spotify.com/v1/tracks/0y6TO3w7SWXO...   
886  https://api.spotify.com/v1/tracks/3o9Yhvfzz2M7...   
887  https://api.spotify.com/v1/tracks/4ghT12J2hsyf...   

                                          analysis_url  duration_ms  \
0    https://api.spotify.com/v1/audio-analysis/2UW7...       234353   
1    https://api.spotify.com/v1/audio-analysis/7ro0...       231704   
2    https://api.spotify.com/v1/audio-analysis/5TML...       179940   
3    https://api.spotify.com/v1/audio-analysis/0Be7...       189320   
4    https://api.spotify.com/v1/audio-analysis/7Fbr...       174253   
..                                                 ...          ...   
883  https://api.spotify.com/v1/audio-analysis/4Rrq...       159387   
884  https://api.spotify.com/v1/audio-analysis/492x...       175173   
885  https://api.spotify.com/v1/audio-analysis/0y6T...       146707   
886  https://ap

In [51]:
# Check for duplicates

audio_features_songs.duplicated().value_counts()

False    883
True       5
dtype: int64

In [53]:
audio_features_songs[audio_features_songs.duplicated()]

danceability  energy  key  loudness  mode  speechiness  acousticness  \
737         0.561   0.965    7    -3.673     0       0.0343       0.00383   
738         0.704   0.797    0    -5.927     1       0.0475       0.08260   
740         0.671   0.845   11    -4.930     0       0.0480       0.02070   
756         0.912   0.774    0    -4.673     0       0.0455       0.12100   
775         0.756   0.697    8    -6.377     1       0.0401       0.18200   

     instrumentalness  liveness  valence    tempo            type  \
737          0.000007    0.3710    0.304  128.040  audio_features   
738          0.000745    0.0546    0.825  139.994  audio_features   
740          0.000000    0.3290    0.775  110.056  audio_features   
756          0.012700    0.1010    0.632  119.966  audio_features   
775          0.000000    0.3330    0.956   94.996  audio_features   

                         id                                   uri  \
737  4uUG5RXrOk84mYEfFvj3cK  spotify:track:4uUG5RXrOk84mYEfFvj3cK   
738  4h9wh7iOZ0GGn8QVp4RAOB  spotify:track:4h9wh7iOZ0GGn8QVp4RAOB   
740  1vYXt7VSjH9JIM5oRRo7vA  spotify:track:1vYXt7VSjH9JIM5oRRo7vA   
756  10xV5l9nhLvFpR8mqzs0bL  spotify:track:10xV5l9nhLvFpR8mqzs0bL   
775  0QBzMgT7NIeoCYy3sJCof1  spotify:track:0QBzMgT7NIeoCYy3sJCof1   

                                            track_href  \
737  https://api.spotify.com/v1/tracks/4uUG5RXrOk84...   
738  https://api.spotify.com/v1/tracks/4h9wh7iOZ0GG...   
740  https://api.spotify.com/v1/tracks/1vYXt7VSjH9J...   
756  https://api.spotify.com/v1/tracks/10xV5l9nhLvF...   
775  https://api.spotify.com/v1/tracks/0QBzMgT7NIeo...   

                                          analysis_url  duration_ms  \
737  https://api.spotify.com/v1/audio-analysis/4uUG...       175238   
738  https://api.spotify.com/v1/audio-analysis/4h9w...       148486   
740  https://api.spotify.com/v1/audio-analysis/1vYX...       176579   
756  https://api.spotify.com/v1/audio-analysis/10xV...       155589   
775  https://api.spotify.com/v1/audio-analysis/0QBz...       206071   

     time_signature                                    title  \
737               4                          I'm Good (Blue)   
738               4                          I Ain't Worried   
740               4  Dance The Night (From Barbie The Album)   
756               4                                     Numb   
775               4               Bam Bam (feat. Ed Sheeran)   

                         artist  
737    David Guetta, Bebe Rexha  
738                 OneRepublic  
740                    Dua Lipa  
756          Marshmello, Khalid  
775  Camila Cabello, Ed Sheeran

In [54]:
audio_features_songs = audio_features_songs.drop_duplicates()
audio_features_songs.duplicated().value_counts()

False    883
dtype: int64

In [55]:
# Drop the columns that we don't need

audio_features_songs = audio_features_songs.drop(["type", "uri", "track_href", "analysis_url"], axis=1)

In [57]:
# Reorder columns in order display the song id, title and artist at the beginning

audio_features_songs = audio_features_songs[['id', 'title', 'artist', 'danceability', 'energy', 'key', 
                                             'loudness', 'mode', 'speechiness', 'acousticness', 
                                             'instrumentalness', 'liveness', 'valence', 'tempo', 
                                             'duration_ms', 'time_signature']]
audio_features_songs.head()

id              title                 artist  \
0  2UW7JaomAMuX9pZrjVpHAU    La Bebe - Remix  Yng Lvcas, Peso Pluma   
1  7ro0hRteUMfnOioTFI5TG1     WHERE SHE GOES              Bad Bunny   
2  5TMLavqGRKNxgzMBrVpC9R  Mon Amour - Remix         zzoilo, Aitana   
3  0Be7sopyKMv8Y8npsUkax2      Tacones Rojos        Sebastian Yatra   
4  7FbrGaHYVDmfr7KoLIZnQ7  Cupid - Twin Ver.            FIFTY FIFTY   

   danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.812   0.479    2    -5.678     0       0.3330        0.2130   
1         0.652   0.800    9    -4.019     0       0.0614        0.1430   
2         0.748   0.761    2    -6.621     1       0.1020        0.1220   
3         0.748   0.856   11    -3.517     0       0.0348        0.0824   
4         0.783   0.592   11    -8.332     0       0.0331        0.4350   

   instrumentalness  liveness  valence    tempo  duration_ms  time_signature  
0          0.000001    0.0756    0.559  169.922       234353               4  
1          0.629000    0.1120    0.234  143.978       231704               4  
2          0.000007    0.0984    0.362  116.041       179940               4  
3          0.000000    0.1420    0.927  123.031       189320               4  
4          0.000004    0.3470    0.726  120.018       174253               4

In [58]:
# Save the csv

audio_features_songs.to_csv("audio_features_songs.csv")

### Conclusion
We obtained a csv with 883 unique songs from different styles (pop, latina, jazz, samba, soul, lofi, country, movie soundtracks...) and moods (happy, sad, romance), with their audio features.